File: P163 Missing Ranges/Solution.java
class Solution {
    // Time: 0(n)
    // Space: O(n)
    public List<String> findMissingRanges(int[] nums, int lower, int upper) {
        List<String> result = new ArrayList<>();
        // All we care about are the ranges of numbers between the lower and upped bounds that do not exist in nums
        int prev = lower - 1;
        for (int i = 0; i <= nums.length; i++) {
            int curr = upper + 1;
            if (i < nums.length) {
                curr = nums[i];
            }
            int start = prev + 1;
            int end = curr - 1;
            if (start <= end) {
                if (start == end) {
                    result.add(String.valueOf(start));
                } else {
                    result.add(start + "->" + end);
                }
            }
            prev = curr;
        }
        return result;
    }
}

File: P163 Missing Ranges/solution.py
class Solution:
    # Time: O(n)
    # Space: O(n)
    def findMissingRanges(self, nums: List[int], lower: int, upper: int) -> List[str]:
        result = []
        prev = lower - 1
        for i in range(0, len(nums) + 1):
            curr = nums[i] if i < len(nums) else upper + 1
            start = prev + 1
            end = curr - 1
            if start <= end:
                if start == end:
                    result.append(str(start))
                else:
                    result.append(str(start) + "->" + str(end))
            prev = curr
        return result